torch/utils/data/distributed.py

```diff
@@ -9,21 +9,30 @@ class DistributedSampler(Sampler):
 
     It is especially useful in conjunction with
     :class:`torch.nn.parallel.DistributedDataParallel`. In such case, each
-    process can pass a DistributedSampler instance as a DataLoader sampler,
+    process can pass a `DistributedSampler` instance as a DataLoader sampler,
     and load a subset of the original dataset that is exclusive to it.
 
     .. note::
         Dataset is assumed to be of constant size.
 
     Arguments:
-        dataset: Dataset used for sampling.
-        num_replicas (optional): Number of processes participating in
-            distributed training.
-        rank (optional): Rank of the current process within num_replicas.
-        shuffle (optional): If true (default), sampler will shuffle the indices
+        dataset (Dataset): Dataset used for sampling. It can be `None` if :attr:`length` is specified
+        num_replicas (int, optional): Number of processes participating in
+            distributed training (default: `None`)
+        rank (int, optional): Rank of the current process within :attr:`num_replicas` (default: `None`)
+        shuffle (bool, optional): If `True` sampler will shuffle the indices (default: `True`)
+        length (int, optional): length of `dataset`
+            If `None`, length is calculated as `len(dataset)` (default: `None`)
+            Must be greater than 0 when :attr:`dataset` is a `IterableDataset`
+        padding (bool, optional): If `True`, the returned lists will be padded to have the same length
+            Padding is done by adding (duplicate) indices from the beggining of the index list
+            into the end of it in a circular fashion. (default: `True`)
+
+    .. note::
+        Regardless of :attr:`padding` value, :meth:`__len__` will return `ceil(dataset length / num_replicas)`
     """
 
-    def __init__(self, dataset, num_replicas=None, rank=None, shuffle=True):
+    def __init__(self, dataset, num_replicas=None, rank=None, shuffle=True, length=None, padding=True):
```

Comment:
`IterableDataset` can still implement `__len__`. I think it makes more sense for this sampler to assume that the dataset has `__len__` than having an explicit input argument.

Reply:
@SsnL if the dataset can implement `len` then it makes more sense to extend `Dataset` as opposed to `IterableDataset`

This PR tries to add the ability to do *distributed* training when the number of samples are *unknown*. `IterableDataset` allows this concept, but for distributed training, the sampler needs some hints on how many chunks this unknown number of samples are split at